src/resolution/instantiation.ts

```diff
@@ -33,39 +32,70 @@ function _createInstance(Func: interfaces.Newable<any>, injections: Object[]): a
     return new Func(...injections);
 }
 
-function _postConstruct(constr: interfaces.Newable<any>, result: any): void {
+function _postConstruct(constr: interfaces.Newable<any>, result: any): void | Promise<void> {
     if (Reflect.hasMetadata(METADATA_KEY.POST_CONSTRUCT, constr)) {
         const data: Metadata = Reflect.getMetadata(METADATA_KEY.POST_CONSTRUCT, constr);
         try {
-            result[data.value]();
+            return result[data.value]();
         } catch (e) {
             throw new Error(POST_CONSTRUCT_ERROR(constr.name, e.message));
         }
     }
 }
 
 function resolveInstance(
+    binding: interfaces.Binding<any>,
     constr: interfaces.Newable<any>,
     childRequests: interfaces.Request[],
     resolveRequest: interfaces.ResolveRequestHandler
 ): any {
+    if (binding.scope === "Transient") {
+        if (typeof binding.onDeactivation === "function") {
+            throw new Error(ON_DEACTIVATION_ERROR(constr.name, "Class cannot be instantiated in transient scope."));
+        }
+
+        if (Reflect.hasMetadata(METADATA_KEY.PRE_DESTROY, constr)) {
+            throw new Error(PRE_DESTROY_ERROR(constr.name, "Class cannot be instantiated in transient scope."));
+        }
+    }
 
     let result: any = null;
 
     if (childRequests.length > 0) {
-
         const constructorInjectionsRequests = childRequests.filter((childRequest: interfaces.Request) =>
             (childRequest.target !== null && childRequest.target.type === TargetTypeEnum.ConstructorArgument));
 
         const constructorInjections = constructorInjectionsRequests.map(resolveRequest);
 
+        if (constructorInjections.some((ci) => ci instanceof Promise)) {
```

Comment:
This isn't that massive a change huh ?

Reply:
nope :) i hacked up half the project the first time around trying to get this to work, and then realized it would be as simple as promise detection.

Reply:
Ponder points:
1.  3rd party Promise libs. Do they generally just inherit from Prototype ? I never use them, but I guess some people (still?) do.
2. What happens when you have "concurrent" async requests going on at once for a given container (Say a parent container, and per http request child containers accessing the parent container for 'shared' deps). Is that supported ? Does this do the right thing in terms of honoring the scopes etc ?

Reply:
1. i'd assume if anyone is using a 3rd party lib, they are probably doing so as a polyfill, which would override window.Promise, and this should still be true.

2. from what I understand of my latest implementation, if there is a race condition with async, it would also exist with non-async (eg: dynamic values, etc). the same semantics of how things are looked up in parent containers and/or cached on bindings ring true with promises.

Reply:
I see, thanks :) So the promises themselves are cached if I am understanding correctly?